lightning/src/ln/peer_handler.rs

```diff
@@ -1483,29 +1553,15 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 					MessageSendEvent::SendChannelUpdate { ref node_id, ref msg } => {
 						log_trace!(self.logger, "Handling SendChannelUpdate event in peer_handler for node {} for channel {}",
 								log_pubkey!(node_id), msg.contents.short_channel_id);
-						let peer = get_peer_for_forwarding!(node_id);
-						peer.pending_outbound_buffer.push_back(peer.channel_encryptor.encrypt_message(&encode_msg!(msg)));
+						self.enqueue_message(&mut *get_peer_for_forwarding!(node_id), msg);
 					},
 					MessageSendEvent::HandleError { ref node_id, ref action } => {
 						match *action {
 							msgs::ErrorAction::DisconnectPeer { ref msg } => {
-								if let Some(mut descriptor) = peers.node_id_to_descriptor.remove(node_id) {
-									if let Some(mut peer) = peers.peers.remove(&descriptor) {
-										if let Some(ref msg) = *msg {
-											log_trace!(self.logger, "Handling DisconnectPeer HandleError event in peer_handler for node {} with message {}",
-													log_pubkey!(node_id),
-													msg.data);
-											self.enqueue_message(&mut peer, msg);
-											// This isn't guaranteed to work, but if there is enough free
-											// room in the send buffer, put the error message there...
-											self.do_attempt_write_data(&mut descriptor, &mut peer);
-										} else {
-											log_gossip!(self.logger, "Handling DisconnectPeer HandleError event in peer_handler for node {} with no message", log_pubkey!(node_id));
-										}
-									}
-									descriptor.disconnect_socket();
-									self.message_handler.chan_handler.peer_disconnected(&node_id, false);
-								}
+								// We do not have the peers write lock, so we just store that we're
+								// about to disconenct the peer and do it after we finish
+								// processing most messages.
+								disconnect_peers.insert(*node_id, msg.clone());
```

Comment:
peers_to_disconnect or peers_pending_disconnection or something like that is a more accurate variable name